base_action_visibility_restriction_domain/models/ir_actions.py

```diff
@@ -0,0 +1,165 @@
+# Copyright 2023 Ooops404 - Ilyas
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+from odoo import _, api, fields, models, tools
+from odoo.exceptions import AccessError
+from odoo.tools.safe_eval import safe_eval
+
+
+class IrActions(models.Model):
+    _inherit = "ir.actions.actions"
+
+    @api.model
+    @tools.ormcache("frozenset(self.env.user.groups_id.ids)", "model_name")
+    def get_bindings(self, model_name):
+        """
+        Filter out actions for which user group is excluded
+        """
+        res = super().get_bindings(model_name)
+        group_ids = set(self.env.user.groups_id.ids)
+        for binding_type, actions in res.items():
+            if binding_type != "action":
+                continue
+            filtered_actions = []
+            for act in actions:
+                act_rec = self.env[act["type"]].browse(act["id"])
+                restrictions = act_rec.restriction_ids.filtered(
+                    lambda x: x.group_id.id in group_ids
+                )
+                if restrictions and any(restrictions.mapped("condition_domain")):
+                    # if at least one has domain then action is visible
+                    # and execution access will be defined in run()
+                    # otherwise action is hidden.
+                    filtered_actions.append(act)
+                elif restrictions and not any(restrictions.mapped("condition_domain")):
+                    # all restrictions are without domain. Action inaccessible.
+                    continue
+                else:
+                    # no restriction. Action accessible.
+                    filtered_actions.append(act)
+            res[binding_type] = filtered_actions
+        return res
+
+
+class IrActionsServer(models.Model):
+    _inherit = "ir.actions.server"
+
+    restriction_ids = fields.One2many("ir.actions.restriction", "server_action_id")
+
+    def run(self):
+        """
+        Prevent action running based on groups and domain
+        """
+        for action in self.sudo():
+            restrictions = action.restriction_ids.filtered(
+                lambda x: x.group_id.id in self.env.user.groups_id.ids
```

Comment:
**suggestion**: like in `get_bindings`, we can move `self.env.user.groups_id.ids` before the `for` loop so it's only accessed once.

Also in this case, we could keep the recordset objects instead of the ids, but that's just preference.
`lambda x: x.group_id in self.env.user.groups_id`